src/contracts/libraries/GPv2Order.sol

```diff
@@ -6,6 +6,133 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+    /// ```
+    /// keccak256(
+    ///     "Order(" +
+    ///         "address sellToken," +
+    ///         "address buyToken," +
+    ///         "address receiver," +
+    ///         "uint256 sellAmount," +
+    ///         "uint256 buyAmount," +
+    ///         "uint32 validTo," +
+    ///         "bytes32 appData," +
+    ///         "uint256 feeAmount," +
+    ///         "string kind," +
+    ///         "bool partiallyFillable" +
+    ///     ")"
+    /// )
+    /// ```
+    bytes32 internal constant TYPE_HASH =
+        hex"d604be04a8c6d2df582ec82eba9b65ce714008acbf9122dd95e499569c8f1a80";
+
+    /// @dev The marker value for a sell order for computing the order struct
+    /// hash. This allows the EIP-712 compatible wallets to display a
+    /// descriptive string for the order kind (instead of 0 or 1).
+    ///
+    /// This value is pre-computed from the following expression:
+    /// ```
+    /// keccak256("sell")
+    /// ```
+    bytes32 internal constant SELL =
+        hex"f3b277728b3fee749481eb3e0b3b48980dbbab78658fc419025cb16eee346775";
+
+    /// @dev The OrderKind marker value for a buy order for computing the order
+    /// struct hash.
+    ///
+    /// This value is pre-computed from the following expression:
+    /// ```
+    /// keccak256("buy")
+    /// ```
+    bytes32 internal constant BUY =
+        hex"6ed88e868af0a1983e3886d5f3e95a2fafbd6c3450bc229e27342283dc429ccc";
+
+    /// @dev The byte length of an order unique identifier.
+    uint256 internal constant UID_LENGTH = 56;
+
+    /// @dev Return the EIP-712 struct hash for the specified order.
+    ///
+    /// @param order The order to compute the EIP-712 struct hash for.
+    /// @return orderDigest The 32 byte EIP-712 struct hash.
+    function hash(Data memory order)
+        internal
+        pure
+        returns (bytes32 orderDigest)
+    {
+        // NOTE: Compute the EIP-712 order struct hash in place. The hash is
+        // computed from the order type hash concatenated with the ABI
+        // encoded order fields for a total of `11 * sizeof(uint) = 352`
+        // bytes. Fortunately, since Solidity memory structs **are not**
+        // packed, they are already laid out in memory exactly as is needed
+        // to compute the struct hash, just requiring the order type hash to
+        // be temporarily written to the memory slot coming right before the
+        // order data.
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            let dataStart := sub(order, 32)
+            let temp := mload(dataStart)
+            mstore(dataStart, TYPE_HASH)
+            orderDigest := keccak256(dataStart, 352)
+            mstore(dataStart, temp)
+        }
+    }
+
+    /// @dev Packs order UID parameters into the specified memory location. The
+    /// result is equivalent to `abi.encodePacked(...)` with the difference that
+    /// it allows re-using the memory for packing the order UID.
```

Comment:
I think it's important to have a comment saying that you must allocate `orderUid` first before using this function, otherwise you'd be writing on weird locations in memory.

Reply:
Fixed.